### ISSUE TITLE
Removing scripts from objects can violate gdscript static typing

### ISSUE THREAD

**Issue by wareya** [godotengine/godot#82448]
### Godot version

4.1.1

### System information

Windows 10, amd rx 480 gpu

### Issue description

If you have a statically typed variable of an instance of a user-defined gdscript class, and then assign null over its script property, then it the instance is no longer of that type. This violates static typing guarantees and will make optimizations around custom class types harder in the future. (For example, `myobj.property` could never optimize down to a single chase-pointers-then-access-array operation, even if myobj is statically typed, because the script might be removed or replaced with a script or another type.)

This will break compatibility, but IMO, set_script() and assigning to the script property of statically-typed `Object`s should throw errors.

### Steps to reproduce

Run a scene with this script attached:

```gdscript
extends Node2D

class MyClass extends RefCounted:
    var x = 0.0

func _init():
    var myobj : MyClass = MyClass.new()
    print(myobj is MyClass)
    myobj.script = null
    print(myobj is MyClass)
```

### Minimal reproduction project

N/A

**Comment by hikari-no-yume** [comment]
Possibly the neatest solution is if using a static type with a scripted object sets some flag on it that prevents detaching of the script later. I believe it'd still be possible to attach scripts with static typing, just not detach them, so it's not too bad.

**Comment by dalexeev** [comment]
I think this is expected. This is not the only place where a user can shoot themselves in the foot. But this is almost impossible to do by accident. I don't think we should do anything other than document it.

> ```gdscript
> myobj.script = null
> ```

This is a bug. The documentation explicitly states that only the setter and getter should be accessible:

> Note: The script is not exposed like most properties. To set or get an object's Script in code, use set_script and get_script, respectively.

**Comment by wareya** [comment]
> I think this is expected. This is not the only place where a user can shoot themselves in the foot. But this is almost impossible to do by accident. I don't think we should do anything other than document it.

My main concern is that this behavior limits future static typing optimizations. If a static Object variable can be of any script type then any external access to its properties will always have to go through the process of checking what script is attached.

> This is a bug. The documentation explicitly states that only the setter and getter should be accessible:

It just defers itself to the getter/setter either way: https://github.com/godotengine/godot/blob/4c3dc26367518e006f8555c12f5d2df0b8a28192/core/object/object.cpp#L352-L353

If anything I would take that point as a documentation bug.

**Comment by Mickeon** [comment]
> If anything I would take that point as a documentation bug.

Not necessarily. You cannot access `script` onto the script itself. Any attempt would generate a "Identifier "script" not declared in the current scope" error. However, you cannot create a variable called "script", because it overshadows it, even though it is hidden and for internal use only.

**Comment by wareya** [comment]
I was reminded of this by seeing https://github.com/godotengine/godot-proposals/issues/15491 and thinking about statically typed optimizations for complex objects again.

It might work to have a way for object types to specify that they forbid script reassignment on existing instances of themselves. Then you don't have to care about whether there are any statically typed bindings referencing that object or whatever else, the type can express whether it's safe to assume that a statically typed bound object is actually that type. Somehow marking their own script property as final or something, for example (I know that that doesn't entirely make sense but it's the right general idea). It also doesn't break backwards compatibility.

**Comment by neikeq** [comment]
> I was reminded of this by seeing godotengine/godot-proposals#15491 and thinking about statically typed optimizations for complex objects again.

I think godotengine/godot-proposals#15451 is even more related to this issue than that one is.

**Comment by wareya** [comment]
Oh, very nice!
